### ISSUE TITLE
LLVM 16 Support

### ISSUE THREAD

**Issue by frantisekz** [intel/intel-graphics-compiler#289]
I understand this isn't a priority, but I'll at least try to create a sort of tracking issue/help where I can (not much tbh).

Distributions are already migrating to LLVM 16 (upcoming Fedora 38 to be released in a few days uses that), and it'll be great if there was planned igc support for that.

Currently hit issues:

- Ton of opencl-clang problems (worked around, created PRs upstream, got it building)
- CCLANG_SONAME_VERSION check doesn't pass (hits <= 5.0.0 path) for opencl-clang built against llvm 16 - disabled that check to be able to progress
- llvm:None was deprecated (advisory to use std::nullopt_t, seems like a drop-in? I can go ahead and create a PR for that)
- error: cannot convert 'llvm::StringLiteral' to 'const char* const' in initialization (hit in IGC/Options/include/igc/Options/)
- probably many more...

I hope @AGindinson wouldn't mind for a ping here? I remember you've worked on fresh LLVM support for past few releases.

**Comment by frantisekz** [comment]
Thanks for the info!

Please, ping here whenever there is something I can help test/provide feedback/etc.

Thanks a lot!

**Comment by eero-t** [comment]
`vc-intrinsics` v0.13.0 (tagged today) has some LLVM 16 support fix: https://github.com/intel/vc-intrinsics/compare/v0.12.3...v0.13.0

**Comment by ConiKost** [comment]
Does anyone know the current status about LLVM 16 support?

**Comment by frantisekz** [comment]
> Does anyone know the current status about LLVM 16 support?

Afaik, it's the same as what I'd reported when creating the ticket (minus Ton of opencl-clang problems and 
CCLANG_SONAME_VERSION check doesn't pass which got resolved opencl-clang side).

**Comment by AGindinson** [comment]
@ConiKost, now that LLVM 14 support has finally reached production quality (and stands as the recommended version, as noted by @frantisekz), we're slowly returning to pursue LLVM 16 compatibility. One of the key issues here is proper opaque pointers' support, which is being investigated in parallel.

**Comment by eero-t** [comment]
Could v16 be added to the LLVM status tracking page: https://github.com/intel/intel-graphics-compiler/projects/5 ?

**Comment by ConiKost** [comment]
@AGindinson thanks for the update!

**Comment by AGindinson** [comment]
> Could v16 be added to the LLVM status tracking page: https://github.com/intel/intel-graphics-compiler/projects/5 ?

@eero-t Done, thanks for the suggestion!

**Comment by ConiKost** [comment]
Any news on LLVM16 support?

**Comment by tjaalton** [comment]
or llvm17 for that matter

**Comment by pszymich** [comment]
Hi, let me shed some light and transparency on the current LLVM situation. Our aim is to start working on LLVM 15 support this quarter with production switch later this year. We are still working on LLVM 14 support for our Windows driver which is the main reason for lack of progress on the open-source side of things.
LLVM updates after switching to 15 will require significant work on opaque pointers situation which is going to majorly affect the roadmap.

**Comment by frantisekz** [comment]
Thanks for the update/confirmation @pszymich .

For Fedora packaging (can't speak about RHEL), I'd switched the igc package to llvm15 which is going to work fine for Fedora for the foreseeable future (and it's still a far better situation compared to eg. intel cm-compiler which has the latest tagged release for llvm 8).

Just (you're probably aware) llvm 16 should work with typed pointers too ( https://llvm.org/docs/OpaquePointers.html#version-support ) if that best effort support works for you and should you need a newer target.

**Comment by pszymich** [comment]
While this is an option - "best effort" is not necessarily something we would be comfortable with in production environment, especially when talking about the foundation of the compiler.
Moving forwards we may consider running testing on such configuration, but any signs of LLVM issues would disqualify 16 integration before opaque pointers support.

**Comment by eero-t** [comment]
> Just (you're probably aware) llvm 16 should work with typed pointers too ( https://llvm.org/docs/OpaquePointers.html#version-support ) if that best effort support works for you and should you need a newer target.

FOSDEM presentation about LLVM in IGC gives info on that: https://fosdem.org/2024/schedule/event/fosdem-2024-2501-challenges-of-supporting-multiple-versions-of-llvm-in-intel-graphics-compiler/

**Comment by tjaalton** [comment]
note that llvm-14 (and i-g-c etc) have been removed from Debian testing, so it won't be in the next release.

**Comment by eero-t** [comment]
While at it, also LLVM version status tracking (https://github.com/intel/intel-graphics-compiler/projects/5) should be moved to new GH projects as GH is phasing out the legacy projects by August.

**Comment by tuliom** [comment]
> note that llvm-14 (and i-g-c etc) have been removed from Debian testing, so it won't be in the next release.

I've just started a discussion in Fedora in order to stop distributing llvm 14, 15 and 16 in Rawhide and Fedora 41.
Link to the discussion: https://lists.fedoraproject.org/archives/list/devel@lists.fedoraproject.org/thread/TDSWXZU27YVCVEBV4UUX7Z7FE56ZFUL5/

**Comment by ConiKost** [comment]
Same applies for Gentoo. LLVM-14 already removed, LLVM-15 also waits for removing and only waits because of IGC. LLVM16 is planed also to be removed in future, since LLVM17 and LLVM18 is already in Gentoo tree.

**Comment by pszymich** [comment]
We are committed to bringing IGC to latest LLVM versions, and we are implementing a much more robust LLVM upgrade processes to accelerate new versions integration. Until new processes are in place, we are also working on LLVM 16 build compatibility in parallel.
We are expecting IGC to be LLVM 16 compatible mid-August, according to experience from previous versions. Also as in previous efforts, we will follow this with fixes to functional/performance quality of the compiler which should happen much faster given new processes.

**Comment by pszymich** [comment]
As mid-August approaches it is most likely we not going to meet the date. Current realistic estimate is mid-September.

**Comment by ConiKost** [comment]
@pszymich Could you please share an update of the status? Thanks!

**Comment by pszymich** [comment]
We've pushed a number of changes addressing LLVM 16 compatibility which resolved around 90% of the raw complier errors. In the next two weeks we should manage to be fully compatible with LLVM 16 APIs.
To build IGC however, we will need to resolve LIT testing errors afterwards, for which we cannot give any reasonable ETA until we see the number of failures.

**Comment by kloczek** [comment]
May I ask ticket owner to change subject to "Latest LLVM 19 Support"?

**Comment by pszymich** [comment]
Once compatible with LLVM 16, we will follow up with a separate ticket for LLVM 17.

**Comment by frantisekz** [comment]
Just a note, in case any of the commenters here (or in the closed issue) are Fedora users wanting to get rid of llvm15 - expect at least half year up to year wait (till F43) as compute-runtime dropped support for pre-Xe architectures, I'll be halting rebases of both igc and compute-runtime to give users on older platforms some support window for their compute workloads.

(Sorry for bit of tangent/off-topic here)

**Comment by ArchangeGabriel** [comment]
@frantisekz I haven’t looked in depth, but it seems you can still build legacy platforms support with `NEO_LEGACY_PLATFORMS_SUPPORT=True`. I will look into it when I have time (which is not anytime soon), but I guess at worst the idea will be to provide two packages, one being stuck at legacy platform last releases.

**Comment by tjaalton** [comment]
llvm-16 is slowly getting removed from Debian already, so I hope the jump to llvm-19 won't be as big as from 14 to 16..
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1075944

**Comment by ConiKost** [comment]
@pszymich Any news?

**Comment by pszymich** [comment]
Since commit https://github.com/intel/intel-graphics-compiler/commit/be3d47298e7f87889b5d8b40e3dcc75e3ebfd07a scalar compiler can be compiled successfully. Vector compiler is still being worked on. Needed adjustments in VC are significant, eg. https://github.com/intel/intel-graphics-compiler/commit/96a6d67b3ce0a311c4226ba835080d732e94c1fe.

Today we are adding LLVM 16 build to the Github Action workflow in this repository.

**Comment by ConiKost** [comment]
@pszymich Thanks for the update. Any ETA for VC?

**Comment by tjaalton** [comment]
will IGC1 ever get proper support for modern llvm?

if not, there's no chance to provide IGC/compute for legacy platforms at least in Debian based distros

**Comment by ConiKost** [comment]
@pszymich ping

**Comment by ConiKost** [comment]
@pszymich Thank you for the update. Are any plans to port back to legacy branch?

**Comment by kloczek** [comment]
> With the [1e48dc5](https://github.com/intel/intel-graphics-compiler/commit/1e48dc5090563f569b0285a08a78ac077dbacdd0) commit all LIT tests should be passing with IGC + LLVM 16. Thank you [@bokrzesi](https://github.com/bokrzesi) for your contributions.

Latest LLVM version it is 19.x.

**Comment by yktyagi** [comment]
> With the [1e48dc5](https://github.com/intel/intel-graphics-compiler/commit/1e48dc5090563f569b0285a08a78ac077dbacdd0) commit all LIT tests should be passing with IGC + LLVM 16. Thank you [@bokrzesi](https://github.com/bokrzesi) for your contributions.
> 
> In related news, we are getting closed to switching production builds to LLVM 15. We should be able to provide more information in coming weeks.

Which version of SPIRV-LLVM-Translator to be used?

**Comment by eero-t** [comment]
> Which version/commit of SPIRV-LLVM-Translator and vc-intrinsics to be used?

While e.g. Ubuntu build doc is _completely_ out of date, release notes list with which versions of the dependencies they've been built with.  That specified combo "should" work.

However, I'm having trouble building latest releases with _system_ LLVM + libs though, regardless of which LLVM version I try (14,15,16...), so I'm wondering whether you're seeing the same problem as I am...

If system SPIRV-LLVM-Translator (version matching to specified LLVM version) is used, IGC  build fails to:
```
IGC/VectorCompiler/lib/Driver/SPIRVWrapper.cpp: In function 'int {anonymous}::spirvReadVerify(const char*, size_t, const uint32_t*, const uint64_t*, unsigned int, llvm::LLVMContext&, void (*)(const char*, size_t, void*), void*, void (*)(const char*, void*), void*)':
#15 180.5 /home/nobody/source/intel-graphics-compiler/IGC/VectorCompiler/lib/Driver/SPIRVWrapper.cpp:117:10: error: 'class SPIRV::TranslatorOpts' has no member named 'setEmitFunctionPtrAddrSpace'
```

But if I tell IGC to use SPIRV-LLVM-Translator _sources_ (from branch matching the specified LLVM version), `compute-runtime` build fails to issue that looks like dep LLVM version mismatch:
```
CommandLine Error: Option 'spirv-expand-step' registered more than once!
#22 39.97 LLVM ERROR: inconsistency in registered CommandLine options
```
 => Seems like IGC would have used system's _default_ LLVM version (18) to build the dep, instead of the specified system LLVM version (e.g. 14)?

**Comment by yktyagi** [comment]
> > Which version/commit of SPIRV-LLVM-Translator and vc-intrinsics to be used?
> 
> While e.g. Ubuntu build doc is _completely_ out of date, release notes list with which versions of the dependencies they've been built with. That specified combo "should" work.
> 
> However, I'm having trouble building latest releases with _system_ LLVM + libs though, regardless of which LLVM version I try (14,15,16...), so I'm wondering whether you're seeing the same problem as I am...
> 
> If system SPIRV-LLVM-Translator (version matching to specified LLVM version) is used, IGC v2 build fails to:
> 
> ```
> IGC/VectorCompiler/lib/Driver/SPIRVWrapper.cpp: In function 'int {anonymous}::spirvReadVerify(const char*, size_t, const uint32_t*, const uint64_t*, unsigned int, llvm::LLVMContext&, void (*)(const char*, size_t, void*), void*, void (*)(const char*, void*), void*)':
> #15 180.5 /home/nobody/source/intel-graphics-compiler/IGC/VectorCompiler/lib/Driver/SPIRVWrapper.cpp:117:10: error: 'class SPIRV::TranslatorOpts' has no member named 'setEmitFunctionPtrAddrSpace'

> ```
> 
> But if I tell IGC v2 to use SPIRV-LLVM-Translator _sources_ (from branch matching the specified LLVM version), `compute-runtime` build fails to issue that looks like dep LLVM version mismatch:
> 
> ```
> CommandLine Error: Option 'spirv-expand-step' registered more than once!
> #22 39.97 LLVM ERROR: inconsistency in registered CommandLine options
> ```
> 
> => Seems like IGC v2 would have used system's _default_ LLVM version (18) to build the dep, instead of the specified system LLVM version (e.g. 14)?
> 
> EDIT: Whereas latest IGC v1 builds fine with system "SPIRV-LLVM-Translator", but supports only LLVM up to v15.

Yes, I also have this same issue.

**Comment by ConiKost** [comment]
@pszymich An update would be really appreciated.

* Will ever IGCv1 receive any modern LLVM support? (I suspect not, as it will stuck forever on LLVM15?)
* What is the state of LLVM16+ for IGCv2?

Comiling [v2.10.8](https://github.com/intel/intel-graphics-compiler/releases/tag/v2.10.8) still fails for me with LLVM16, but works with LLVM15.

```
FAILED: IGC/VectorCompiler/lib/Driver/CMakeFiles/VCDriver.dir/Driver.cpp.o
/usr/lib/ccache/bin/x86_64-pc-linux-gnu-g++ -DBIF_LINK_BC -DCL_KHR_FP64_EXT -DGHAL3D=USC -DICBE_LINUX -DIGA_DISABLE_ENCODER_EXCEPTIONS -DIGC_CMAKE -DIGC_EXPORTS=1 -DIGC_SPIRV_ENABLED -DIGC_SPIRV_TOOLS_ENABLED -DIGC_VC_ENABLED -DINSIDE_PLUGIN -DISTDLIB_UMD -DLINUX -DLLVM_VERSION_MAJOR=16 -DNDEBUG -DNOMINMAX -DSTD_CALL -DUSC_EXPORTS=1 -DUSE_MMX -DUSE_SSE -DUSE_SSE2 -DUSE_SSE3 -DUSE_SSSE3 -D_AMD64_ -D_COMPILER_DLL_ -D_CRT_SECURE_NO_WARNINGS -D_HAS_EXCEPTIONS=0 -D_IGC_ -D_SCL_SECURE_NO_WARNINGS -D__IGC_OPAQUE_POINTERS_API_ENABLED=true -D__IGC_OPAQUE_POINTERS_DEFAULT_ARG_CLANG=\"-opaque-pointers\" -D__STDC_CONSTANT_MACROS -D__STDC_LIMIT_MACROS -I/var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9/IGC/WrapperLLVM/include -I/var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9_build/IGC/autogen -I/var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9/IGC -I/var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9/IGC/common -I/var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9/IGC/.. -I/var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9/IGC/../Common -I/var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9/IGC/../3d/common -I/var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9/IGC/../inc -I/var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9/IGC/../inc/common -I/var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9/IGC/../inc/common/Compiler -I/var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9/IGC/../inc/common/Compiler/API -I/var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9/IGC/../visa/include -I/var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9_build/IGC/Release -I/var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9_build/IGC -I/var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9/IGC/AdaptorOCL/ocl_igc_shared/executable_format -I/var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9/IGC/AdaptorOCL -I/var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9_build/IGC/AdaptorOCL/ocl_igc_interface/impl -I/var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9/IGC/Options -I/var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9/IGC/Options/include -I/var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9_build/IGC/Options/include -I/var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9/IGC/../inc/common/Compiler/common -I/var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9/IGC/AdaptorOCL/cif/cif/.. -I/var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9/IGC/../../../SPIRV-LLVM-Translator/include -I/var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9/IGC/../../../SPIRV-LLVM-Translator/lib/SPIRV -I/var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9/IGC/../../../SPIRV-LLVM-Translator/lib/SPIRV/Mangler -I/var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9/IGC/ZEBinWriter/zebin/source -I/var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9/IGC/ZEBinWriter/zebin/source/autogen -I/include -I/var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9/IGC/VectorCompiler/lib/Driver/../GenXCodeGen -I/var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9_build/IGC/VectorCompiler/lib/Driver/../../lib/GenXCodeGen -I/var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9/IGC/VectorCompiler/include -I/var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9_build/IGC/VectorCompiler/include -I/var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9_build/IGC/VectorCompiler/include/vc/InternalIntrinsics -isystem /usr/lib/llvm/16/include -isystem /usr/lib/llvm/16/include/LLVMSPIRVLib  -DNDEBUG -march=native -mfpmath=sse,387 -mtune=intel -frecord-gcc-switches -fomit-frame-pointer -malign-data=abi -mtls-dialect=gnu2 -pipe -fvisibility-inlines-hidden -U_GLIBCXX_ASSERTIONS -fexceptions -fdata-sections -ffunction-sections -O2 -fmessage-length=0 -march=corei7 -mstackrealign -fms-extensions -Wno-unused-parameter -Wno-missing-field-initializers -Wwrite-strings -Wno-long-long -Wswitch -Wno-sign-compare -Wno-unused-result -Wno-enum-compare -Wno-type-limits -Wno-ignored-qualifiers -Wno-shadow -Wformat -Wformat-security -Wno-extra -Wno-write-strings -finline -fno-strict-aliasing -msse -msse2 -msse3 -mssse3 -msse4 -msse4.1 -msse4.2 -Wno-unknown-pragmas -fPIC -Bsymbolic -fstack-protector -finline-functions -funswitch-loops -Wno-maybe-uninitialized -lrt -fno-rtti -fvisibility=hidden -g -std=gnu++17 -MD -MT IGC/VectorCompiler/lib/Driver/CMakeFiles/VCDriver.dir/Driver.cpp.o -MF IGC/VectorCompiler/lib/Driver/CMakeFiles/VCDriver.dir/Driver.cpp.o.d -o IGC/VectorCompiler/lib/Driver/CMakeFiles/VCDriver.dir/Driver.cpp.o -c /var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9/IGC/VectorCompiler/lib/Driver/Driver.cpp
In file included from /usr/lib/llvm/16/include/llvm/Support/Threading.h:21,
                 from /usr/lib/llvm/16/include/llvm/Support/RWMutex.h:17,
                 from /usr/lib/llvm/16/include/llvm/PassRegistry.h:23,
                 from /usr/lib/llvm/16/include/llvm/PassSupport.h:29,
                 from /usr/lib/llvm/16/include/llvm/Pass.h:348,
                 from /usr/lib/llvm/16/include/llvm/Analysis/TargetTransformInfo.h:28,
                 from /var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9/IGC/WrapperLLVM/include/llvmWrapper/Analysis/TargetTransformInfo.h:13,
                 from /var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9/IGC/VectorCompiler/lib/Driver/../GenXCodeGen/GenXTargetMachine.h:21,
                 from /var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9/IGC/VectorCompiler/lib/Driver/Driver.cpp:13:
/usr/lib/gcc/x86_64-pc-linux-gnu/15/include/g++-v15/ciso646:46:4: warning: #warning "<ciso646> is deprecated in C++17, use <version> to detect implementation-specific macros" [-Wcpp]
   46 | #  warning "<ciso646> is deprecated in C++17, use <version> to detect implementation-specific macros"
      |    ^~~~~~~
In file included from /var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9/IGC/VectorCompiler/lib/Driver/Driver.cpp:61:
/var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9/IGC/WrapperLLVM/include/llvmWrapper/IR/LLVMContext.h: In function ‘bool IGC::canOverwriteLLVMCtxPtrMode(llvm::LLVMContext*)’:
/var/tmp/portage/dev-util/intel-graphics-compiler-2.10.9/work/intel-graphics-compiler-2.10.9/IGC/WrapperLLVM/include/llvmWrapper/IR/LLVMContext.h:49:22: error: ‘class llvm::LLVMContext’ has no member named ‘hasSetOpaquePointersValue’; did you mean ‘setOpaquePointers’?
   49 |         return !Ctx->hasSetOpaquePointersValue();
      |                      ^~~~~~~~~~~~~~~~~~~~~~~~~
      |                      setOpaquePointers
ninja: build stopped: subcommand failed.
```

**Comment by ConiKost** [comment]
Well, it seems, that you are applying this patch inhouse?
```external/llvm/releases/16.0.0/patches_external/Reapply_hasSetOpaquePointersValue.patch```

I don't see this as an (valid) option, to include it in distros LLVM16.

**Comment by tjaalton** [comment]
The oldest llvm available on the next Debian release is 17.0.6, and it's not going to be around for long in the devel series. So the target should preferably be 19+ at this point.

**Comment by zboszor** [comment]
> The oldest llvm available on the next Debian release is 17.0.6, and it's not going to be around for long in the devel series. So the target should preferably be 19+ at this point.

Should be, but isn't.

For one, Yocto 5.2 has just been released with LLVM/CLANG 20.1.x. Intel's own [meta-intel](https://git.yoctoproject.org/meta-intel) does not handle this properly:

- `opencl-clang` is still using version 15.0.0 from branch `ocl-open-150` from its source repository while there is also the `ocl-open-200` branch with version 20.1.0.
- `intel-graphics-compiler` version 2.10.10 does not build against LLVM 20.1.2 headers.

I am slowly picking up the garbage in my own [meta-clang-revival](https://github.com/zboszor/meta-clang-revival/) Yocto layer. It seems I will have to add recipe addendums to build the non-legacy IGC version against LLVM 15.

**Comment by zboszor** [comment]
At this point, Intel should rather spend time improving Rusticl and Iris or ANV+Zink to support missing OpenCL features compared to the Intel Compute Runtime. Mesa already supports LLVM 20/21.

**Comment by basiliscos** [comment]
>  Intel should rather spend

I think they are in bad financial state, so I'd not expect big advances from their side..

**Comment by ConiKost** [comment]
> My previous comment is incorrect. IGC 2.18 still uses LLVM 15. IGC 2.19 will use LLVM 16 with typed pointers. I apologize for the confusion made.

@pszymich Could you please clarify, if that means, that IGC 2.19 will need LLVM16 compiled with the patch ```external/llvm/releases/16.0.0/patches_external/Reapply_hasSetOpaquePointersValue.patch``` applied for now?

**Comment by eero-t** [comment]
> The oldest llvm available on the next Debian release is 17.0.6, and it's not going to be around for long in the devel series. So the target should preferably be 19+ at this point.

@tjaalton On the Ubuntu side, another issue is the other packages built with LLVM 15, that IGC needs, being too old for IGC v2.

For example some SPIRV-LLVM-Translator APIs needed by IGC v2 were added in v15.0.6 (with latest of that branch being currently v15.0.16), but latest Ubuntu is still shipping only v15.0.1 (and v1 IGC): https://packages.ubuntu.com/source/plucky/spirv-llvm-translator-15

**Comment by smunaut** [comment]
It can also be re-titled "LLVM 21 Support" :)

**Comment by Anutrix** [comment]
I found https://packages.debian.org/forky/intel-opencl-icd with LLVM16(?) (but no page for trixie).

Does this mean we can hope Debian 14(Forky) will have intel-opencl-icd ?

**Comment by tjaalton** [comment]
that version in Debian vendors llvm-16 inside it (needed for ubuntu 26.04), but it's not meant to be released with forky
